includes/core/server/class-request.php

```diff
@@ -0,0 +1,523 @@
+use WCPay\Core\Exceptions\Immutable_Parameter_Exception;
+use WCPay\Core\Exceptions\Invalid_Request_Parameter_Exception;
+
+/**
+ * Base for requests to the WCPay server.
+ */
+abstract class Request {
+	/**
+	 * Contains a set of params, which the class considers immutable by others.
+	 *
+	 * Overwrite this in your class for individual properties.
+	 *
+	 * @var string[]
+	 */
+	const IMMUTABLE_PARAMS = [];
+
+	/**
+	 * Indicates which parameters are required (keys only).
+	 *
+	 * @var string[]
+	 */
+	const REQUIRED_PARAMS = [];
+
+	/**
+	 * Contains default values for parameters, which are not set automatically.
+	 *
+	 * @var string[]
+	 */
+	const DEFAULT_PARAMS = [];
+
+	/**
+	 * Holds the parameters of the request.
+	 *
+	 * @var []
+	 */
+	private $params = [];
+
+	/**
+	 * True when `->apply_filters()` is called to protect read-only props.
+	 *
+	 * In protected mode, if somebody tries to change an immutable param,
+	 * as declared in `IMMUTABLE_PARAMS`, an exception will be thrown.
+	 *
+	 * This way important params can be safe from modifications by extensions.
+	 *
+	 * @var bool
+	 */
+	private $protected_mode = false;
+
+	/**
+	 * Holds the API client of WCPay.
+	 *
+	 * @var WC_Payments_API_Client
+	 */
+	protected $api_client;
+
+	/**
+	 * Holds the HTTP interface of WCPay.
+	 *
+	 * @var WC_Payments_Http_Interface
+	 */
+	protected $http_interface;
+
+	/**
+	 * Prevents the class from being constructed directly.
+	 *
+	 * @param WC_Payments_API_Client     $api_client     The API client to use to send requests.
+	 * @param WC_Payments_Http_Interface $http_interface The HTTP interface for the server.
+	 */
+	public function __construct( WC_Payments_API_Client $api_client, WC_Payments_Http_Interface $http_interface ) {
+		$this->api_client     = $api_client;
+		$this->http_interface = $http_interface;
+	}
+
+	/**
+	 * Returns the needed API.
+	 *
+	 * @return string Check WCPay\Core\Server\APIs.
+	 */
+	abstract public function get_api(): string;
+
+	/**
+	 * Returns the method of the request.
+	 *
+	 * @return string See the constants in WordPress's `Requests` class.
+	 */
+	abstract public function get_method(): string;
+
+	/**
+	 * This is a legacy method, and is the same throughout the codebase.
+	 * Might be worth removing while refactoring to use the Core\Server API.
+	 *
+	 * @return bool
+	 */
+	public function is_site_specific(): bool {
+		return true;
+	}
+
+	/**
+	 * If true, the request will be signed with the user token rather than blog token. Defaults to false.
+	 *
+	 * @return bool
+	 */
+	public function should_use_user_token(): bool {
+		return false;
+	}
+
+	/**
+	 * Indicates if the raw response should be returned.
+	 *
+	 * @return bool
+	 */
+	public function should_return_raw_response(): bool {
+		return false;
+	}
+
+	/**
+	 * Returns all of the parameters for the request.
+	 *
+	 * @return array
+	 * @throws Invalid_Request_Parameter_Exception If the request has not been initialized yet.
```

Comment:
```suggestion
* @throws Invalid_Request_Parameter_Exception If the request has not been initialized properly.
```

It's not clear what "initialization" means in this context